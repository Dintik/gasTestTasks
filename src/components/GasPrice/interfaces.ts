export interface IPrice {
    gasPrice: string,
    ethRate: string
}