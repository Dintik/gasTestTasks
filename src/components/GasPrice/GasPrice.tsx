import React, { useEffect, useCallback } from "react";
import { useSelector, useDispatch } from "react-redux";
import { setGasPrice, setEthRate } from "../../store/prices/reducer";
import { getStoreGasPrice, getStoreEthRate } from "../../store/prices/helpers";
import styles from "./styles.module.scss";
import { getGasPrice, getEthRate } from "../../helpers";

const GasPrice: React.FC = () => {
    const price = {
        gas: useSelector(getStoreGasPrice),
        eth: useSelector(getStoreEthRate)
    };
    const dispatch = useDispatch();

    const refreshData = useCallback(async function () {
        dispatch(setGasPrice("...")); // you can add for clarity
        dispatch(setEthRate("...")); // you can add for clarity
        const gasPrice = await getGasPrice();
        const ethRate = await getEthRate();
        dispatch(setGasPrice(gasPrice));
        dispatch(setEthRate(ethRate));
    }, []);

    useEffect(() => {
        refreshData();
    }, []);

    return (
        <section className={styles.gasPrice}>
            <button className={styles.gasPrice__btn} type="button" onClick={refreshData}>
                Refresh
            </button>
            <div className={styles.gasPrice__text__wrap}>
                <p className={styles.gasPrice__text}>Gas price: {price?.gas} gwei</p>
                <p className={styles.gasPrice__text}>Exchange rate: {price?.eth}$</p>
            </div>
        </section>
    );
};

export default GasPrice;
