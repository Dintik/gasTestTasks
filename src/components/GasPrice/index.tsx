import React, { useEffect, useCallback } from "react";
import { useAppSelector, useAppDispatch } from "../../store/hooks";
import { setGasPrice, setEthRate, setLoading } from "../../store/prices/reducer";
import styles from "./styles.module.scss";
import { getEthRate } from "../../api/getEthRate";
import { getGasPrice } from "../../web3/utils";
import { IPrice } from "./interfaces";

const GasPrice: React.FC = () => {
    const price: IPrice = useAppSelector((state) => state.pricesState);
    const dispatch = useAppDispatch();

    const refreshData = useCallback(async function () {
        dispatch(setLoading());
        const curentprice: IPrice = {
            gasPrice: await getGasPrice(),
            ethRate: await getEthRate()
        };
        curentprice.gasPrice && dispatch(setGasPrice(curentprice.gasPrice));
        curentprice.ethRate && dispatch(setEthRate(curentprice.ethRate));
    }, []);

    useEffect(() => {
        refreshData();
    }, []);

    return (
        <section className={styles.gasPrice}>
            <button className={styles.gasPrice__btn} type="button" onClick={refreshData}>
                Refresh
            </button>
            <div className={styles.gasPrice__text__wrap}>
                <p className={styles.gasPrice__text}>Gas price: {price.isLoadingEth ? "Loading..." : `${price.gasPrice} gwei`}</p>
                <p className={styles.gasPrice__text}>Exchange rate: {price.isLoadingGas ? "Loading..." : `${price.ethRate}$`}</p>
            </div>
        </section>
    );
};

export default GasPrice;
