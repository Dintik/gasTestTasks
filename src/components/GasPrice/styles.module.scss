.gasPrice {
    text-align: center;
    min-height: 100vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    font-size: calc(10px + 2vmin);

    &__text {
        font-size: calc(10px + 2vmin);
        margin: 20px;
        min-width: 364px;
        text-align: left;

        &__wrap {
            padding-bottom: 150px;
        }
    }

    &__btn {
        font-size: calc(10px + 2vmin);
        margin: 10px;
        padding: 20px 30px;
    }
}
