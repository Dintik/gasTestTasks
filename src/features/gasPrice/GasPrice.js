import React, { useEffect, useCallback, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { selectGasPrice, setGasPrice } from "./gasPriceGet";

export function GasPrice() {
    const { ethers } = require("ethers");
    const gasPrice = useSelector(selectGasPrice);
    const dispatch = useDispatch();

    const [gasPriceUsd, setGasPriceUsd] = useState("...");

    const getGasPrice = useCallback(async function () {
        dispatch(setGasPrice("..."));
        const provider = new ethers.providers.Web3Provider(window.ethereum);
        const feeData = await provider.getFeeData();
        const gasPrice = feeData.gasPrice;
        const gasPriceString = ethers.utils.formatUnits(gasPrice, 9);
        const gasPriceWei = Number(gasPriceString).toFixed(3);
        dispatch(setGasPrice(gasPriceWei));

        // fetch("https://api.coinbase.com/v2/prices/ETH-USD/buy")
        //     .then((response) => {
        //         return response.json();
        //     })
        //     .then((data) => {
        //         console.log(data);
        //     });

        fetch(
            "https://api.coingecko.com/api/v3/simple/price?ids=ethereum&vs_currencies=usd"
        )
            .then((response) => {
                return response.json();
            })
            .then((data) => {
                setGasPriceUsd(data?.ethereum?.usd);
            });
    }, []);

    useEffect(() => {
        getGasPrice();
    }, []);

    return (
        <div className="App-header-wrap">
            <div>
                <p className="App-header-text">Gas price: {gasPrice} gwei</p>
                <p className="App-header-text">Exchange rate: {gasPriceUsd}$</p>
            </div>
            <button
                className="App-header-btn"
                type="button"
                onClick={getGasPrice}
            >
                Refresh
            </button>
        </div>
    );
}
