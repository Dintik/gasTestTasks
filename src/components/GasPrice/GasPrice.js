import React, { useEffect, useCallback, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { setGasPrice } from "../../store/gasPrice/reducer";
import { getStoreGasPrice } from "../../store/gasPrice/helpers";
import styles from "./styles.module.scss";
import { getGasPrice, getEthRate } from "../../helpers";

const GasPrice = () => {
    const gasPrice = useSelector(getStoreGasPrice);
    const dispatch = useDispatch();
    const [ethPrice, setEthPrice] = useState(0);

    const refreshData = useCallback(async function () {
        dispatch(setGasPrice("..."));
        setEthPrice("...");
        const gasPrice = await getGasPrice();
        const ethRate = await getEthRate();
        dispatch(setGasPrice(gasPrice));
        setEthPrice(ethRate);
    }, []);

    useEffect(() => {
        refreshData();
    }, []);

    return (
        <section className={styles.gasPrice}>
            <div>
                <p className={styles.gasPrice__text}>
                    Gas price: {gasPrice} gwei
                </p>
                <p className={styles.gasPrice__text}>
                    Exchange rate: {ethPrice}$
                </p>
            </div>
            <button
                className={styles.gasPrice__btn}
                type="button"
                onClick={refreshData}
            >
                Refresh
            </button>
        </section>
    );
};

export default GasPrice;
