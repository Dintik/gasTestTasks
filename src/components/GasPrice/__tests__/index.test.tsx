import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import GasPrice from "../index";
const mockStore = configureMockStore([thunk]);

describe("Gas Price", () => {
    it("should render loading", () => {
        const store = mockStore({
            pricesState: {
                gasPrice: null,
                ethRate: null,
                isLoadingGas: true,
                isLoadingEth: true
            }
        });

        const { asFragment } = render(
            <Provider store={store}>
                <GasPrice />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render correctly without data", () => {
        const store = mockStore({
            pricesState: {
                ethRate: "Sorry, there was an error",
                gasPrice: "Sorry, there was an error"
            }
        });

        const { asFragment } = render(
            <Provider store={store}>
                <GasPrice />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render data", () => {
        const store = mockStore({
            pricesState: {
                ethRate: "2697.51",
                gasPrice: "154.788"
            }
        });

        const { asFragment } = render(
            <Provider store={store}>
                <GasPrice />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});

export {};
