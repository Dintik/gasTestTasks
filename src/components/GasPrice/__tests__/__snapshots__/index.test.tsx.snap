// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gas Price should render correctly without data 1`] = `
<DocumentFragment>
  <section
    class="gasPrice"
  >
    <button
      class="gasPrice__btn"
      type="button"
    >
      Refresh
    </button>
    <div
      class="gasPrice__text__wrap"
    >
      <p
        class="gasPrice__text"
      >
        Gas price: Sorry, there was an error gwei
      </p>
      <p
        class="gasPrice__text"
      >
        Exchange rate: Sorry, there was an error$
      </p>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Gas Price should render data 1`] = `
<DocumentFragment>
  <section
    class="gasPrice"
  >
    <button
      class="gasPrice__btn"
      type="button"
    >
      Refresh
    </button>
    <div
      class="gasPrice__text__wrap"
    >
      <p
        class="gasPrice__text"
      >
        Gas price: 154.788 gwei
      </p>
      <p
        class="gasPrice__text"
      >
        Exchange rate: 2697.51$
      </p>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Gas Price should render loading 1`] = `
<DocumentFragment>
  <section
    class="gasPrice"
  >
    <button
      class="gasPrice__btn"
      type="button"
    >
      Refresh
    </button>
    <div
      class="gasPrice__text__wrap"
    >
      <p
        class="gasPrice__text"
      >
        Gas price: Loading...
      </p>
      <p
        class="gasPrice__text"
      >
        Exchange rate: Loading...
      </p>
    </div>
  </section>
</DocumentFragment>
`;
